### PR TITLE
Refactor regularization step closes #117

### PR DESCRIPTION
**Pull request by SebChw**
To see this step in action please check

```
 python -m art.cli regularization-tutorial
```

[get rid off turn_on/off regularization](https://github.com/SebChw/Actually-Robust-Training/commit/6521d24ffb80e985760044a67c5f7f09019aab4a) - We won't use it for regularization

[add datamodule modifiers](https://github.com/SebChw/Actually-Robust-Training/commit/930a8de24b1a6385e45415f60b5ecbff3d38befe) - regularization will be based on these modifiers

[reduce dependencies](https://github.com/SebChw/Actually-Robust-Training/commit/848f4a1a71bf3b83afd9afed8a56e403d25fde30) - since we use them just once they are not needed

[add access to current stage for ArtModule](https://github.com/SebChw/Actually-Robust-Training/commit/87adae356b55e688d00cc797b8ecb6640abb12e9) - During regularization when augmentations are used I must use them only during training

[add utils for orange logs](https://github.com/SebChw/Actually-Robust-Training/commit/3177ae1aa32ddf1a8ad3ba4646ec181ac0919057) - I made some warning orange

[add utils for building metric name](https://github.com/SebChw/Actually-Robust-Training/commit/1ef112f40517eabd928e6fb986e0895239f1b871) - Small util usefull when we want to define ModelCallback

[add possibility for the step to decide when it should be run or not](https://github.com/SebChw/Actually-Robust-Training/commit/cc0a0ffffc78f2dcbd8ee5567c83b3c0131877b5) - This is a big deal as for me - new regularizations should be continued even though the previous one failed to obtain good accuracy. That's why selected steps can decide if they should be run or not. I know this is not with SOLID but I couldn't find any other way

[separate model and data logs](https://github.com/SebChw/Actually-Robust-Training/commit/ad90ba621e18a11efbf2cf7125cdd64908776914) - When log_params were executed on dataloader withouth `setup` there was a crush.

[rewrite regularizaton](https://github.com/SebChw/Actually-Robust-Training/commit/b964dec7661ed525ac1c3d95d4064b83600b65b2) - Rewritten step

[add class fields](https://github.com/SebChw/Actually-Robust-Training/commit/522c858a7ad236971fd64999b866b0d124f27901) - Decide which steps can decide for themselves

[adjust dashboard](https://github.com/SebChw/Actually-Robust-Training/commit/a221c60275e4b4c05b44990832b351aa3172ebd2) - Since regularize steps adds new field to parameters dashboard must handle it

### CODE REVIEW

File: art/metrics.py [hunk 1]
```diff
@@ -31,6 +31,17 @@ def __init__(
         self.stages = stages
 
 
+def build_metric_name(metric: Any, stage: str) -> str:
```
Comment (kordc) [inline art/metrics.py:34]: Nice, but why is it outside of any class?

File: art/steps.py [hunk 2]
```diff
@@ -609,11 +631,68 @@ def do(self, previous_states: Dict):
...
+            return
+
+        # Load all previous runs and check if any of them has the same parameters
+        runs = JSONStepSaver().load(self.get_full_step_name())["runs"]
+        for run in runs:
+            if run["parameters"]["regularize"] == self.parameters:
+                self.was_already_tried = True
+                self.results = run
+
+    def check_if_already_tried(self):
```
Comment (kordc) [inline art/steps.py:647]: why do you separate the "update" and the "check" of already tried? I think it should be one function, even if it returns something in Python it isn't a problem

File: art/steps.py [hunk 3]
```diff
@@ -14,8 +14,14 @@
 
 from art.core import ArtModule
 from art.decorators import ModelDecorator, art_decorate
-from art.loggers import (add_logger, art_logger, get_new_log_file_name,
-                         get_run_id, log_yellow_warning, remove_logger)
+from art.loggers import (
```
Comment (kordc) [inline art/steps.py]: Here you unfold the imports you've already folded in https://github.com/SebChw/Actually-Robust-Training/pull/200/commits/2e7b764d43a12f25ab36c2e209581b0f2f730eac

Comment (SebChw) [reply art/steps.py]: Well, it's automatic formatting thing... I didn't mean to do so

Comment (SebChw) [reply art/steps.py]: Maybe formatter got off on the wrong foot

File: art/metrics.py (hunk 1 above)
Comment (SebChw) [reply art/metrics.py:34]: As this is just a utility function. I want to make it easily usable by the user.

File: art/steps.py (hunk 2 above)
Comment (SebChw) [reply art/steps.py:647]: Actually Idk. Probably I was using this method in some other place too and it turned out not to be necessary.

Comment (SebChw) [reply art/steps.py:647]: But probably my logic was that checking is not about updating and this self.was-already_tried is used in few places.

**Review by kordc — APPROVED** [review]

**Pull request merged**